internal/indexer/postgres/affectedmanifest.go

```diff
@@ -0,0 +1,248 @@
+	// and determine if any in-tree matcher finds the record vulnerable
+	var filteredRecords []claircore.IndexRecord
+	om := omnimatcher.New(nil)
+	for _, pkg := range pkgsToFilter {
+		pr.Package = &pkg
+		if om.Vulnerable(&pr, &v) {
+			p := pkg // make a copy, or else you'll get a stale reference later
+			filteredRecords = append(filteredRecords, claircore.IndexRecord{
+				Package:      &p,
+				Distribution: pr.Distribution,
+				Repository:   pr.Repository,
+			})
+		}
+	}
+	log.Debug().Int("count", len(filteredRecords)).Msg("vulnerable indexrecords")
+
+	// query the manifest index for manifests containing
+	// the vulnerable indexrecords and create a set
+	// containing each unique manifest
+	set := map[string]claircore.Digest{}
+	var buf bytes.Buffer
+	for _, record := range filteredRecords {
+		buf.Reset()
+		err := affectedQuery.Execute(&buf, record)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate sql query for IndexRecord %+v: %v", record, err)
+		}
+		rows, err := pool.Query(ctx,
+			buf.String(),
+			record.Package.ID,
+		)
+		if err != nil {
+			rows.Close()
+			if err == pgx.ErrNoRows {
+				continue
+			}
+			return nil, fmt.Errorf("failed to query the manifest index: %v", err)
+		}
+		for rows.Next() {
+			var hash claircore.Digest
+			err := rows.Scan(&hash)
+			if err != nil {
+				rows.Close()
+				return nil, fmt.Errorf("failed scanning manifest hash into digest: %v", err)
+			}
+			set[hash.String()] = hash
+		}
+	}
+
+	n := len(set)
+	out := make([]claircore.Digest, 0, n)
+	for _, v := range set {
+		out = append(out, v)
+	}
+	log.Debug().Int("count", len(out)).Msg("affected manifests")
+	return out, nil
+}
+
+// protoRecord is a helper method which resolves a Vulnerability to an IndexRecord with no Package defined.
+//
+// it is an error for both a distribution and a repo to be missing from the Vulnerability.
+func protoRecord(ctx context.Context, pool *pgxpool.Pool, v claircore.Vulnerability) (claircore.IndexRecord, error) {
+	const (
+		selectDist = `
+		SELECT id
+		FROM dist
+		WHERE arch = $1
+		  AND cpe = $2
+		  AND did = $3
+		  AND name = $4
+		  AND pretty_name = $5
+		  AND version = $6
+		  AND version_code_name = $7
+		  AND version_id = $8;
+		`
+		selectRepo = `
+		SELECT id
+		FROM repo
+		WHERE name = $1
+			AND key = $2
+			AND uri = $3;
+		`
+	)
+	log := zerolog.Ctx(ctx).
+		With().
+		Str("component", "internal/indexer/postgres/protoRecord").
+		Logger()
+
+	protoRecord := claircore.IndexRecord{}
+	// fill dist into prototype index record if exists
+	if (v.Dist != nil) && (v.Dist.Name != "") {
+		row := pool.QueryRow(ctx,
+			selectDist,
+			v.Dist.Arch,
+			v.Dist.CPE,
+			v.Dist.DID,
+			v.Dist.Name,
+			v.Dist.PrettyName,
+			v.Dist.Version,
+			v.Dist.VersionCodeName,
+			v.Dist.VersionID,
+		)
+		var id pgtype.Int8
+		err := row.Scan(&id)
+		if err != nil {
+			if err != pgx.ErrNoRows {
+				return protoRecord, fmt.Errorf("failed to scan dist: %v", err)
+			}
+		}
```

Comment:
This should use `errors.Is` to make sure it's all unwrapped correctly:
```suggestion
		switch {
		case err == nil:
		case errors.Is(err, pgx.ErrNoRows): // OK
		default:
			return protoRecord, fmt.Errorf("failed to scan dist: %v", err)
		}
```